pkg/controllers/clusterresourceplacementeviction/controller.go

```diff
@@ -0,0 +1,257 @@
+)
+
+// Reconciler reconciles a ClusterResourcePlacementEviction object.
+type Reconciler struct {
+	client.Client
+}
+
+// Reconcile triggers a single eviction reconcile round.
+func (r *Reconciler) Reconcile(ctx context.Context, req runtime.Request) (runtime.Result, error) {
+	startTime := time.Now()
+	evictionName := req.NamespacedName.Name
+	klog.V(2).InfoS("ClusterResourcePlacementEviction reconciliation starts", "clusterResourcePlacementEviction", evictionName)
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("ClusterResourcePlacementEviction reconciliation ends", "clusterResourcePlacementEviction", evictionName, "latency", latency)
+	}()
+
+	var eviction placementv1alpha1.ClusterResourcePlacementEviction
+	if err := r.Client.Get(ctx, req.NamespacedName, &eviction); err != nil {
+		klog.ErrorS(err, "Failed to get cluster resource placement eviction", "clusterResourcePlacementEviction", evictionName)
+		return runtime.Result{}, client.IgnoreNotFound(err)
+	}
+
+	validCondition := eviction.GetCondition(string(placementv1alpha1.PlacementEvictionConditionTypeValid))
+	if condition.IsConditionStatusFalse(validCondition, eviction.GetGeneration()) {
+		klog.V(2).InfoS("Invalid eviction, no need to reconcile", "clusterResourcePlacementEviction", evictionName)
+		return runtime.Result{}, nil
+	}
+
+	executedCondition := eviction.GetCondition(string(placementv1alpha1.PlacementEvictionConditionTypeExecuted))
+	if executedCondition != nil {
+		klog.V(2).InfoS("Eviction has executed condition specified, no need to reconcile", "clusterResourcePlacementEviction", evictionName)
+		return runtime.Result{}, nil
+	}
+
+	isCRPPresent := true
+	var crp placementv1beta1.ClusterResourcePlacement
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: eviction.Spec.PlacementName}, &crp); err != nil {
+		if !errors.IsNotFound(err) {
+			return runtime.Result{}, err
+		}
+		isCRPPresent = false
+	}
+	if !isCRPPresent {
+		klog.V(2).InfoS(evictionInvalidMissingCRP, "clusterResourcePlacementEviction", evictionName, "clusterResourcePlacement", eviction.Spec.PlacementName)
+		markEvictionInvalid(&eviction, evictionInvalidMissingCRP)
+		return runtime.Result{}, r.updateEvictionStatus(ctx, &eviction)
+	}
+
+	var crbList placementv1beta1.ClusterResourceBindingList
+	if err := r.Client.List(ctx, &crbList, client.MatchingLabels{placementv1beta1.CRPTrackingLabel: crp.Name}); err != nil {
+		return runtime.Result{}, err
+	}
+
+	var evictionTargetBinding *placementv1beta1.ClusterResourceBinding
+	for i := range crbList.Items {
+		if crbList.Items[i].Spec.TargetCluster == eviction.Spec.ClusterName {
+			if evictionTargetBinding == nil {
+				evictionTargetBinding = &crbList.Items[i]
+			} else {
+				klog.V(2).InfoS(evictionInvalidMultipleCRB, "clusterResourcePlacementEviction", evictionName, "clusterResourcePlacement", eviction.Spec.PlacementName)
```

Comment:
Currently when an eviction is performed, CRB watcher watches the deletion and triggers a scheduling cycle. Then the scheduler adds deleting binding in any state to a deleting list and removes finalizers.

Due to this the scheduler will currently create another CRB for the same cluster, since the old binding is accounted for in the deleting binding list.